# Data Analysis With Residential Homes

In [1]:
import numpy as np
import pandas as pd

In [2]:
pd.set_option('display.width',90)

In [3]:
#Reading "homes" data set containing information about 522 residential houses. 
df = pd.read_csv('homes.csv')
df[:5]

price  area  beds  baths  garage  year  style  lotsize   ac pool quality highway
0  360000  3032     4      4       2  1972      1    22221  YES   NO  MEDIUM      NO
1  340000  2058     4      2       2  1976      1    22912  YES   NO  MEDIUM      NO
2  250000  1780     4      3       2  1980      1    21345  YES   NO  MEDIUM      NO
3  205500  1638     4      2       2  1963      1    17342  YES   NO  MEDIUM      NO
4  275500  2196     4      3       2  1968      7    21786  YES   NO  MEDIUM      NO

#### Task 1: Find the number of houses from each style.

In [4]:
#Investigating the type of data.
df.shape, df.dtypes

((522, 12),
 price       int64
 area        int64
 beds        int64
 baths       int64
 garage      int64
 year        int64
 style       int64
 lotsize     int64
 ac         object
 pool       object
 quality    object
 highway    object
 dtype: object)

In [5]:
# Converting "style" data type from integer to object (categorical variable).
df['style'] = df['style'].astype(object)
df['style'].dtypes

dtype('O')

In [6]:
df['style'].value_counts()

1     214
7     136
3      64
2      58
5      18
6      18
4      11
9       1
10      1
11      1
Name: style, dtype: int64

The above data show the number of houses from different style which was generated from grouping the data set into category. \
The right column is representing the number of houses with the leeft column as the style of houses.

#### Task 2: Find the smallest, median, and largest value for each numerical column in the dataset.

In [7]:
# DataFrame with the numerical columns
df2 = df.loc[:,df.dtypes != object]
df2.describe().round(2)

price     area    beds   baths  garage     year   lotsize
count     522.00   522.00  522.00  522.00  522.00   522.00    522.00
mean   277894.15  2260.63    3.47    2.64    2.10  1966.90  24369.70
std    137923.40   711.07    1.01    1.06    0.65    17.64  11684.08
min     84000.00   980.00    0.00    0.00    0.00  1885.00   4560.00
25%    180000.00  1701.25    3.00    2.00    2.00  1956.00  17204.75
50%    229900.00  2061.00    3.00    3.00    2.00  1966.00  22200.00
75%    335000.00  2636.25    4.00    3.00    2.00  1981.00  26786.75
max    920000.00  5032.00    7.00    7.00    7.00  1998.00  86830.00

In [8]:
# Selecting rows for min, median, and max values.
list1 = [3,5,7]
df2.describe().iloc[list1]

price    area  beds  baths  garage    year  lotsize
min   84000.0   980.0   0.0    0.0     0.0  1885.0   4560.0
50%  229900.0  2061.0   3.0    3.0     2.0  1966.0  22200.0
max  920000.0  5032.0   7.0    7.0     7.0  1998.0  86830.0

The above data show the smallest, median, and largest value containing numeric data type.

#### Task 3: Find the most expensive house with at least three bedrooms.

In [9]:
df3 = df[df.beds >= 3]
df3[df3.price == df3.price.max()]

price  area  beds  baths  garage  year style  lotsize   ac pool quality highway
72  920000  3857     4      5       3  1997     1    32793  YES   NO    HIGH      NO

The above row is the most expensive house having at least three bedrooms with the price of $920,000. 

#### Task 4: Find all houses with 2-4 beds, area exceeding 4000, price not less than $350000

In [10]:
df4 = df[(df.beds >= 2) & (df.beds <= 4) 
         & (df.area > 4000) 
         & (df.price >= 350000)]
df4

price  area  beds  baths  garage  year style  lotsize   ac pool quality highway
73   855000  4756     4      4       3  1990     7    22215  YES   NO    HIGH      NO
77   665000  4746     4      4       3  1996     7    23368  YES   NO    HIGH      NO
80   780000  4419     4      5       7  1987     1    56127  YES   NO    HIGH      NO
154  395000  4150     4      3       3  1934     7    21778   NO   NO  MEDIUM      NO

From the above data we can find that there are four houses with the price tag of $350,000 or more, 2-4 beds, and the area square footage exceeding 4,000. 

#### Task 5: Find the average lot size of houses built after 1970

In [11]:
# The average lotsize per year
df5 = df[df.year > 1970]
df5.groupby('year').agg([np.mean])['lotsize']

mean
year              
1971  24580.000000
1972  23670.000000
1973  19669.750000
1974  25688.000000
1975  17952.333333
1976  27226.428571
1977  26132.714286
1978  24757.055556
1979  25417.666667
1980  23703.400000
1981  20453.600000
1982  26193.000000
1983  20623.200000
1984  19287.875000
1985  15310.000000
1986  20968.333333
1987  29826.466667
1988  20922.857143
1989  23792.000000
1990  20567.500000
1991  28611.000000
1992  31373.222222
1993  15599.333333
1994  23473.333333
1995  28701.000000
1996  28060.250000
1997  26571.714286
1998  21999.000000

The above data show the average lot size of houses built after 1970.  

#### Task 6: On average, how much more expensive are houses with a pool?

In [12]:
df6 = df.groupby('pool').agg([np.mean])['price']
df6

mean
pool               
NO    272395.915638
YES   352120.277778

In [13]:
df6.iloc[1].round(2)-df6.iloc[0].round(2)

mean    79724.36
dtype: float64

From the above data show that on average the house with a pool is $79,724.36 more expensive.

#### Task 7: Find the smallest, average, and largest price of houses by quality.

In [14]:
df.groupby('quality').agg([min,np.mean,max])['price']

min           mean     max
quality                               
HIGH     325000  543610.647059  920000
LOW       84000  175018.292683  359900
MEDIUM   140400  273766.279310  675000

From the above data we can see the smallest, average, and largest price of houses by high, medium, and low quality.

#### Task 8: Number of houses classified by style and number of beds.

In [15]:
pd.crosstab(df['style'],df.beds)

beds   0  1   2   3   4   5  6  7
style                            
1      1  8  46  92  55  11  1  0
2      0  0   0  31  24   2  1  0
3      0  0   6  34  18   6  0  0
4      0  0   3   5   2   1  0  0
5      0  0   3   4   6   5  0  0
6      0  0   0   7   7   4  0  0
7      0  0   6  28  67  23  9  3
9      0  0   0   1   0   0  0  0
10     0  0   0   0   0   0  1  0
11     0  1   0   0   0   0  0  0

The table above shows the number of houses classified by style and number of beds. 

#### Task 9: Find the least expensive style 1 house.

In [16]:
df9 = df[df['style'] == 1]
df9[df9.price == df9.price.min()]

price  area  beds  baths  garage  year style  lotsize  ac pool quality highway
53  84000   980     1      1       1  1951     1    17686  NO   NO     LOW      NO

The above row show the least expensive style 1 house with the price of $84,000.

#### Task 10: Find number of houses for all categorical columns.

In [17]:
df10 = df.loc[:,df.dtypes == object]
df10[:5]

style   ac pool quality highway
0     1  YES   NO  MEDIUM      NO
1     1  YES   NO  MEDIUM      NO
2     1  YES   NO  MEDIUM      NO
3     1  YES   NO  MEDIUM      NO
4     7  YES   NO  MEDIUM      NO

In [18]:
for c in df10.columns: 
    print("\n",c)
    print(df10[c].value_counts())


 style
1     214
7     136
3      64
2      58
5      18
6      18
4      11
9       1
10      1
11      1
Name: style, dtype: int64

 ac
YES    434
NO      88
Name: ac, dtype: int64

 pool
NO     486
YES     36
Name: pool, dtype: int64

 quality
MEDIUM    290
LOW       164
HIGH       68
Name: quality, dtype: int64

 highway
NO     511
YES     11
Name: highway, dtype: int64


The above data show the number of houses for all categorial columns. 